engine/src/execution_graph/logic_controllers/BatchWindowFunctionProcessing.cpp

```diff
@@ -32,48 +33,93 @@ ComputeWindowKernel::ComputeWindowKernel(std::size_t kernel_id, const std::strin
-                throw std::runtime_error("In Window Function: RANGE is not currently supported");
+    if (is_first_value_window(this->type_aggs_as_str[pos]) || is_last_value_window(this->type_aggs_as_str[pos])) {
+        
+        // When using cudf::grouped_rolling_window() it crashes
+        // rolling_detail.cuh:738: Aggregation operator and/or input type combination is invalid
+        // https://github.com/rapidsai/cudf/issues/7231 expecting unit test using make_nth_element_aggregation()
+
+        // first: we want to get all the first (or last) values (due to the columns to partition)
+        std::vector<cudf::groupby::aggregation_request> requests;
+        requests.emplace_back(cudf::groupby::aggregation_request());
+        requests[0].values = col_view_to_agg;
+        requests[0].aggregations.push_back(std::move(window_aggregation));
+        
+        cudf::groupby::groupby gb_obj(cudf::table_view({table_view_to_partition}), cudf::null_policy::EXCLUDE, cudf::sorted::YES, {}, {});
+        std::pair<std::unique_ptr<cudf::table>, std::vector<cudf::groupby::aggregation_result>> result = gb_obj.aggregate(requests);
+
+        windowed_col = std::move(result.second[0].results[0]);
+
+        // if exists duplicated values (in table_view_to_partition) we want to fill the `windowed_col` by repeating values
+        if (windowed_col->size() < col_view_to_agg.size()) { 
+            std::unique_ptr<cudf::table> table_view_to_partition_wo_repetitions = cudf::drop_duplicates(table_view_to_partition,
+                                                                                             column_partitioned_indices,
+                                                                                             cudf::duplicate_keep_option::KEEP_FIRST);
+
+            // TODO this is just a default setting. Will want to be able to properly set null_order
+            std::vector<cudf::null_order> null_orders(column_partitioned_indices.size(), cudf::null_order::AFTER);
+            std::vector<cudf::order> sortOrderTypes(column_partitioned_indices.size(), cudf::order::ASCENDING);
+            std::unique_ptr<cudf::column> pivot_indexes = cudf::lower_bound(table_view_to_partition,
+                                                                            table_view_to_partition_wo_repetitions->view(),
+                                                                            sortOrderTypes,
+                                                                            null_orders);
+            // we want to get the differences between two contiguous pivots
+            std::vector<cudf::size_type> pivot_indexes_host = ral::utilities::column_to_vector<cudf::size_type>(pivot_indexes->view());
+            std::vector<cudf::size_type> offsets;
+            for (size_t i = 0; i < pivot_indexes_host.size() - 1; ++i) {
+                offsets.push_back(pivot_indexes_host[i + 1] - pivot_indexes_host[i]);
```

Comment:
If you have a large amount of partitions, this offsets vector could be large. This means this operation you are doing here can be slow. Please optimize this. 
See if you can calculate the difference on GPU. If you cant easily do it with cudf, you should be able to use thrust

Reply:
Actually a much better way to do this whole thing is to do a join, which would likely be faster.
For example, you do your group by on the partition keys with the first or last value agregation. Then instead of the lowebound/offsets/repeat thing. You can instead do a join from on the partition keys that have aggregated, to the partition keys that have not been aggregated. 
The thing to be careful there is if the output of the join will have the results out of order than what you want. So test that out first. And you can also see if there is a sort based join, as opposed to a hash based join. I think a sort based join would keep the order the way you want, since you would be sorting the join keys which would be the same as the partition keys which should be sorted already anyways

Reply:
And to clarify, by a join here, i mean just using cudf join operations, locally on the batch being processed, not a full bsql join